### ISSUE TITLE
Compilation can fail with NoClassDefFoundError when low on memory

### ISSUE THREAD

**Issue by eatkins** [sbt/zinc#914]
## steps

In the zinc project, using jdk8, G1GC and a max heap of 1150m run:
`sbt -J-Xmx1150m -Dsbt.version=1.4.0-RC1 -J-XX:+UseG1GC -J-verbose:gc "clean;publishLocal"`
## problem

```
[error] (compilerBridgeJVM2_12 / Compile / compileIncremental) java.lang.NoClassDefFoundError: xsbt/DelegatingReporter$CompileProblem
[error] (compilerBridgeJVM2_10 / Compile / compileIncremental) java.lang.NoClassDefFoundError: xsbt/DelegatingReporter$CompileProblem
[error] (compilerBridgeJVM2_10 / Compile / doc) java.lang.NoClassDefFoundError: xsbt/DelegatingReporter$CompileProblem
[error] (compilerBridgeJVM2_12 / Compile / doc) java.lang.NoClassDefFoundError: xsbt/DelegatingReporter$CompileProblem
[error] (compilerBridgeJVM2_11 / Compile / compileIncremental) java.lang.NoClassDefFoundError: xsbt/ExtractAPI$$anonfun$mkAnnotations$1$$anonfun$apply$1
[error] (compilerBridgeJVM2_11 / Compile / doc) java.lang.NoClassDefFoundError: xsbt/DelegatingReporter$CompileProblem
[error] Total time: 101 s (01:41), completed Sep 8, 2020 11:28:48 AM
```

## expectation

It should have OOMed.

## notes

Bumping the heap to 1250m, the following command works:
`sbt -J-Xmx1250m -Dsbt.version=1.4.0-RC1 -J-XX:+UseG1GC -J-verbose:gc "clean;publishLocal"`

**Comment by eatkins** [comment]
I don't know what, if anything, we can do about this but I wanted to create the issue so that this behavior becomes known. I feel like I have seen weird NoClassDefFoundErrors before and never realized it could be due to insufficient memory.

**Comment by smarter** [comment]
Interesting, this is probably the root cause of https://github.com/sbt/zinc/issues/750, is this a JVM bug?

**Comment by eed3si9n** [comment]
Here's my speculation. AnalyzingCompiler constructs a layered ClassLoader in `loader` method: https://github.com/sbt/zinc/blob/9a64b54fa1f45234b86d0ab175f8c96c348c1b80/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/AnalyzingCompiler.scala#L326-L342

This tries to avoid creating duplicate `ClassLoader`s by first checking `classLoaderCache`, which internally uses SoftReference - https://github.com/sbt/zinc/blob/9a64b54fa1f45234b86d0ab175f8c96c348c1b80/internal/zinc-classpath/src/main/scala/sbt/internal/inc/classpath/ClassLoaderCache.scala#L127

Normally the eviction timing of these ClassLoader is not apparent to the build users, but when the heap is close to the limit, the eviction happens in the middle of multiple `compile` tasks?

**Comment by eatkins** [comment]
That's an interesting theory. The soft reference shouldn't be collected if the classloader is being used anywhere. But it should be easy to add logging to see if that is happening.

**Comment by eatkins** [comment]
Your theory was correct. If I delete the line in the sbt ClassLoaderCache.scala that closes the classloader when it's evicted from the cache due to memory pressure, then publishLocal succeeds.

**Comment by eatkins** [comment]
With #915, I was able to publishLocal with -Xmx750m and G1GC. I did eventually OOM when I dropped it to 600m.

**Comment by eatkins** [comment]
Closing due to #915

**Comment by mvitaly** [comment]
Is it possible that is bug is not resolved ? I'm getting very similar outcome with gradle on large project. Tried to backport the fix to zinc 1.3.5 which is used by gradle (6.8.2) and that didn't help.

I'm trying to understand the fix. If the classloader is evicted from cache and `.close()` is called, another thread that would still be using the classloader for compilation won't be able to load the next class, right ?

**Comment by eatkins** [comment]
IIRC, gradle does not use the sbt classloader cache which is the construct that closes classloaders in sbt when they are no longer needed and they system is under memory pressure, which it detects by maintaining soft references to the classloaders. #915 was added to specifically fix unwanted closes performed by the sbt classloader cache. There was a PR related to the gradle classloader cache which left me with the impression that gradle was not doing the right thing wrt classloader caching compared to sbt: https://github.com/gradle/gradle/pull/12866. Note that the comment after mine suggests that they are using strong references for the cache which seems risky if there are multiple threads involved in the build.
